app/components/blacklight/date_facet_hierarchy_component.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+# Display the date facet in a tree hierarchy
+module Blacklight
+  class DateFacetHierarchyComponent < Blacklight::FacetFieldListComponent
+    def facet_item_presenters
+      facet_item_tree_hierarchy_from_year_list.map do |item|
+        facet_item_presenter(item)
+      end
+    end
+
+    private
+
+    # Build out the tree hierarchy (century -> decade -> year) for
+    # a facet and return the top-level facet items
+    #
+    # @return [Array<Blacklight::Solr::Response::Facets::FacetItem>]
+    def facet_item_tree_hierarchy_from_year_list
+      tree = {}
+
+      @facet_field.paginator.items.each do |item|
+        century_str, decade_str, year_str = item.value.split(':')
+
+        tree[century_str] ||= Blacklight::Solr::Response::Facets::FacetItem.new(value: century_str, items: [], hits: 0)
+
+        next unless decade_str
+
+        decade_facet = tree[century_str].items.find do |decade_item|
```

Comment:
I had a little trouble following when this case would ever be true, because 1) you explicitly set ` tree[century_str].items` to an empty array on line 24, and 2) you've already split the string on `:` on line 22. 

I concluded that this `decade_facet` could only exist if you'd already created it during another loop.

Reply:
yeah, that's right, this code is looking to update an existing instance if it finds one, or to add that decade for the first time if need be.  will add a comment...

Reply:
added some inline comments that try to explain what the conditionals mean in the larger context (of both the current loop iteration and the prior loop iterations), as well as a couple example strings that would meet the condition for skipping to the next loop iteration.